reporting/src/request/steps/token-telemetry/index.js

```diff
@@ -72,111 +71,94 @@ export default class TokenTelemetry {
     this.config = config;
     this.trustedClock = trustedClock;
     this.shouldCheckToken = shouldCheckToken;
-    this.subjectTokens = new Subject();
-    this.tokenSendQueue = new Subject();
-    this.keySendQueue = new Subject();
-
+    this.batch = [];
     this.tokens = new TokenPipeline({
       name: 'tokens',
       db: database.tokens,
       trustedClock,
       options: opts,
+      sendMessage: (payload) =>
+        this.telemetry({ action: 'wtm.attrack.tokensv2', payload }),
+      batchInterval: this.TOKEN_BATCH_INTERVAL,
+      batchLimit: this.TOKEN_BATCH_SIZE,
     });
     this.keys = new KeyPipeline({
       name: 'keys',
       db: database.keys,
       trustedClock,
       options: opts,
+      sendMessage: (payload) =>
+        this.telemetry({ action: 'wtm.attrack.keysv2', payload }),
+      batchInterval: this.KEY_BATCH_INTERVAL,
+      batchLimit: this.KEY_BATCH_SIZE,
     });
   }
 
   async init() {
     await this.tokens.isReady;
     await this.keys.isReady;
-    let filteredTokensBatch = [];
-    const filteredTokens = new Subject();
+    await this.tokens.init();
+    await this.keys.init();
+
     setInterval(() => {
-      filteredTokens.pub(filteredTokensBatch);
-      filteredTokensBatch = [];
+      this.#processBatch([...this.batch]);
+      this.batch = [];
     }, this.TOKEN_BUFFER_TIME);
 
-    this.subjectTokens.subscribe((token) => {
-      filteredTokensBatch.push(token);
-    });
+    // run every x minutes while there is activity
+    setInterval(async () => {
+      await this.tokens.clean();
+      await this.keys.clean();
+    }, this.CLEAN_INTERVAL);
+  }
 
-    // token subscription pipeline takes batches of tokens (grouped by value)
-    // caches their state, and pushes values for sending once they reach a sending
-    // threshold.
+  // token subscription pipeline takes batches of tokens (grouped by value)
+  // caches their state, and pushes values for sending once they reach a sending
+  // threshold.
+  #processBatch(batch) {
+    if (batch.length === 0) {
+      return;
+    }
     const today = this.trustedClock.getTimeAsYYYYMMDD();
```

Comment:
Interesting that "today" used be fixed at startup, as long as the background was alive. Looked unintended because "today" could easily become yesterday, or the week before.

Moving to service workers should have changed that already; so, I would not expect it to have a significant effect now. Still, technically it is a change in the raw data (just a note for us to check in the raw data).